lib/cn/move.c

```diff
@@ -29,54 +29,62 @@ cn_move(
     bool                       src_del,
     struct cn_tree_node       *tgt_node)
 {
-    struct cn_tree *tree;
-    struct kvset_list_entry *src, *src_end, *tgt;
+    struct kvset_list_entry *src, *tgt;
     struct list_head *src_head, *tgt_head;
-    uint64_t *src_ksidv = NULL;
+    struct cn_tree *tree;
     uint64_t src_nodeid;
-    merr_t err;
 
     INVARIANT(w && src_node && tgt_node);
-    INVARIANT((src_cnt > 0 && src_list) || (src_cnt == 0 && !src_list));
+    INVARIANT(!src_cnt == !src_list);
 
     tree = w->cw_tree;
     src_head = &src_node->tn_kvset_list;
     tgt_head = &tgt_node->tn_kvset_list;
 
     src = src_list;
-    src_end = NULL;
     if (src) {
+        uint64_t *src_ksidv;
+        merr_t err;
+        void *lock;
+
         src_ksidv = malloc(sizeof(*src_ksidv) * src_cnt);
         if (!src_ksidv)
             return merr(ENOMEM);
 
+        /* Must hold the tree lock while iterating over the kvset list
```

Comment:
Yes, kvsets can come and go from this list while you're walking it w/o the lock.  So you could end up with src_end being bogus, which is what happend:  When we got into the code below src_end was no longer on the list, and so the loop eventually segfaulted.

Reply:
Right. But my question was after your fix, do we need to hold the tree lock in read mode?

Reply:
Which fix, there are so many!   You are searching forward, so you need the head to be stable in order know if you run into it, and if you don't hit the head then the node you hit could disappear if you're not holding the lock.  So, not sure I understand what you're asking...

Reply:
Sorry for not being clear.

I was referring to your fix which now uses 'src_cnt' in this loop instead of the buggy 'src_end' logic.

'src_list' is pointing to one of the entries in the src_node->tn_kvset_list.
It is on the caller to ensure the linkage for 'src_cnt' items starting from 'src_list' doesn't change when calling this function.
list_next_entry_or_null() on src_cnt - 1 items can be trusted.
list_next_entry_or_null() on the last (src_cnt) entry cannot be trusted without holding the lock but it seems fine to me because we don't use the entry after the last one.
The 'src_head' should not change underneath us as it is just a pointer to a dummy head node (struct list_head embedded inside struct cn_tree_node).
src_head->next and src_head->prev can change underneath us but we don't access those pointers.

There's no harm in taking the read lock here but I was just trying to understand what issue do we run into without holding it as we now loop through only the marked nodes which the caller has already locked for us...

Reply:
The "or_null()" macros do access the head prev/next fields.  IMHO you're playing with fire with this line of thinking when it comes to reasoning about how it oughta work on SMT, a lot of things "just work" on x86 that won't work on other architectures.  You could switch to using the cds list variants which do carefully ordered operations with appropriate barriers, maybe we should!

Reply:
Good idea to switch to cds list variants!

>>> The "or_null()" macros do access the head prev/next fields.
Yes, it does access head->next in list_empty().
list_empty() must always return false in this case because we got entries in the list (src_cnt > 0), i.e., head->next can never be equal to head while we are here.